src/lsstseries/analysis/structure_function/iqr/calculator.py

```diff
@@ -0,0 +1,123 @@
+from typing import List
+
+import numpy as np
+from scipy.stats import binned_statistic
+
+from lsstseries.analysis.structure_function.base_argument_container import StructureFunctionArgumentContainer
+from lsstseries.analysis.structure_function.base_calculator import StructureFunctionCalculator
+
+# For details see Kozlowski 16 Equation 10: https://arxiv.org/abs/1604.05858
+CONVERSION_TO_SIGMA = 0.741
+
+
+class IqrStructureFunctionCalculator(StructureFunctionCalculator):
+    """This class implements the structure function calculation described in
+    Equation 10 of Kozlowski 16: https://arxiv.org/abs/1604.05858
+
+    SF_obs(deltaT) = 0.741 * IQR
+
+    Where `IQR` is the interquartile range between 25% and 75% of the sorted
+    (y(t) - y(t+delta_t)) distribution.
+    """
+
+    def __init__(
+        self,
+        time: List[List[float]],
+        flux: List[List[float]],
+        err: List[List[float]],
+        argument_container: StructureFunctionArgumentContainer,
+    ):
+        super().__init__(time, flux, err, argument_container)
+
+    def calculate(self):
+        all_d_fluxes = []
+        for lc_idx in range(len(self._time)):
+            lc_times = self._time[lc_idx]
+            lc_fluxes = self._flux[lc_idx]
+
+            # mask out any nan values
+            t_mask = np.isnan(lc_times)
+            f_mask = np.isnan(lc_fluxes)
+            lc_mask = np.logical_or(t_mask, f_mask)
+
+            lc_times = lc_times[~lc_mask]
+            lc_fluxes = lc_fluxes[~lc_mask]
+
+            # compute difference of times
+            dt_matrix = lc_times.reshape((1, lc_times.size)) - lc_times.reshape((lc_times.size, 1))
+            d_times = dt_matrix[dt_matrix > 0].flatten()
+
+            # compute difference of fluxes, keep only where difference in time > 0
+            df_matrix = lc_fluxes.reshape((1, lc_fluxes.size)) - lc_fluxes.reshape((lc_fluxes.size, 1))
+            d_fluxes = df_matrix[dt_matrix > 0].flatten()
+
+            # build stack of times and fluxes
+            # `self._dts` and `all_d_fluxes` will have shape = (num_lightcurves, N)
+            self._dts.append(d_times)
+            all_d_fluxes.append(d_fluxes)
+
+        # combining treats all lightcurves as one when calculating the structure function
+        if self._argument_container.combine and len(self._time) > 1:
+            self._dts = np.hstack(np.array(self._dts, dtype="object"))
+            all_d_fluxes = np.hstack(np.array(all_d_fluxes, dtype="object"))
```

Comment:
Just checking that I understand: This concatenates all of the light curve deltas (deltas in both time and flux). But the zero points of each of these light curves will be different. Would it make more sense to concatenate before computing `dt_matrix` and `df_matrix` so the light curves are all baselined to the same zero value?

Reply:
Also chatted with Neven about this - here we want to combine the delta_t and delta_magnitudes, not the observation timestamps and observed magnitudes. So in this case, we're doing what is expected.